### PR TITLE
Api for registering custom languages

### PR DESCRIPTION
**Pull request by RedCMD**
Allows loading a custom wasm file and assigning it to a custom language id
Usage: `registerLanguage(languageId, wasm)`
with `wasm` being a absolute path to the `.wasm` file or the languageId of a builtin `.wasm` file

~~Can overwrite the built in languages with custom wasms~~
Can register a custom language id with one of the built in wasms

~~Automatically reloads all open documents if their language changes~~

https://github.com/cursorless-dev/vscode-parse-tree/issues/34

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -225,6 +232,27 @@ export async function activate(context: vscode.ExtensionContext) {
...
+    registerLanguage(languageId, wasm) {
+      if (languages[languageId] != null) {
+        const language = languages[languageId];
+        if (language.parser != null) {
+          language.parser.delete();
+        }
+      }
+      languages[languageId] = { module: wasm };
+
+      for (const uriString in trees) {
```
Comment (pokey) [inline src/extension.ts:244]:
```suggestion
      for (let uriString in trees) {
```

File: src/extension.ts [hunk 2]
```diff
@@ -225,6 +232,27 @@ export async function activate(context: vscode.ExtensionContext) {
...
+        if (vscode.workspace.textDocuments.find(textDocument =>
```
Comment (pokey) [inline src/extension.ts:245]:
```suggestion
        if (vscode.workspace.textDocuments.some(textDocument =>
```

**Review by pokey — COMMENTED** [review]
Looks good to me! Left a couple minor tweaks. 

Also, would be good to add a blurb to the readme.

**Comment by RedCMD** [comment]
Can register with `registerLanguage(languageId, wasm);`
if the `languageId` already exists. it simply returns nothing (does not overwrite)

File: src/extension.ts [hunk 3]
```diff
@@ -225,6 +232,15 @@ export async function activate(context: vscode.ExtensionContext) {
   return {
     loadLanguage,
 
+    registerLanguage(languageId, wasm) {
+      if (languages[languageId] != null) {
+          return;
```
Comment (pokey) [inline src/extension.ts:237]: I think I'd throw an exception in this case

File: src/extension.ts [hunk 4]
```diff
@@ -225,6 +232,15 @@ export async function activate(context: vscode.ExtensionContext) {
   return {
     loadLanguage,
 
+    registerLanguage(languageId, wasm) {
```
Comment (pokey) [inline src/extension.ts:235]: I think I'd call this `wasmPath` and put a quick note in a docstring to indicate that the path has to be absolute

**Review by pokey — COMMENTED** [review]
Looking good!  Left a couple minor comments

**Comment by RedCMD** [comment]
ping? @pokey

**Review by pokey — APPROVED** [review]
Made a couple tweaks

**Comment by pokey** [comment]
Released in 0.23.0; out in the next 5–10 minutes

**Pull request merged**